arangod/Cluster/Maintenance.cpp

```diff
@@ -511,6 +516,84 @@ VPackBuilder getShardMap(VPackSlice const& collections) {
   return shardMap;
 }
 
+auto arangodb::maintenance::diffReplicatedLogs(
+    DatabaseID const& database, ReplicatedLogStatusMap const& localLogs,
+    ReplicatedLogSpecMap const& planLogs, std::string const& serverId,
+    MaintenanceFeature::errors_t& errors,
+    std::unordered_set<DatabaseID>& makeDirty, bool& callNotify,
+    std::vector<std::shared_ptr<ActionDescription>>& actions) -> arangodb::Result {
+  using namespace arangodb::replication2;
+
+  auto const createReplicatedLogAction = [&](LogId id, agency::LogPlanSpecification const* spec) {
+    auto specStr = std::invoke([&] {
+      VPackBuilder builder;
+      auto slice = VPackSlice::noneSlice();
+      if (spec != nullptr) {
+        spec->toVelocyPack(builder);
+        slice = builder.slice();
+      }
+      return StringUtils::encodeBase64(slice.startAs<char>(), slice.byteSize());
+    });
+
+    auto description = std::make_shared<ActionDescription>(
+        std::map<std::string, std::string>{
+            {std::string(NAME), std::string(UPDATE_REPLICATED_LOG)},
+            {std::string(DATABASE), database},
+            {REPLICATED_LOG_ID, std::to_string(id.id())},
+            {REPLICATED_LOG_SPEC, specStr},
+        },
+        NORMAL_PRIORITY, false);
+
+    makeDirty.insert(database);
+    callNotify = true;
+    actions.emplace_back(std::move(description));
+  };
+
+  // check all plan log entries
+  for (auto const& [logId, spec] : planLogs) {
+    if (spec.currentTerm && spec.currentTerm->participants.find(serverId) !=
+                                spec.currentTerm->participants.end()) {
+      // check if there are logs that do not exist locally
+      if (auto localIt = localLogs.find(spec.id); localIt == std::end(localLogs)) {
+        createReplicatedLogAction(spec.id, &spec);
+      } else {
+        // check if the term is the same
+        bool const requiresUpdate = std::invoke([&, &status = localIt->second, &spec = spec] {
+          if (spec.currentTerm.has_value()) {
+            auto currentTerm = status.getCurrentTerm();
+            return !currentTerm.has_value() || *currentTerm != spec.currentTerm->term;
+          }
+          return false;
+        });
+
+        // Create UpdateLogAction
+        if (requiresUpdate) {
+          createReplicatedLogAction(spec.id, &spec);
+        }
+      }
+    }
+  }
+
+  for (auto const& [id, status] : localLogs) {
+    bool const dropLog = std::invoke([&, &id = id] {
+      auto it = planLogs.find(id);
+      if (it == std::end(planLogs)) {
+        return true;
+      }
+      auto const& spec = it->second;
+      return !spec.currentTerm.has_value() ||
+             (spec.currentTerm->participants.find(serverId) ==
+              spec.currentTerm->participants.end());
+    });
+
+    if (dropLog) {
+      createReplicatedLogAction(id, nullptr);
+    }
+  }
+
+  return Result{};
```

Comment:
The function result will always be an empty `Result` object, so we could declare it void.